### DISCUSSION TITLE
Neqsim Python questions

### DISCUSSION THREAD

**Discussion by andr1976** [equinor/neqsim#467] [General]
Hi 

I am fairly new to NeqSim, and couldn't find a discussion page under neqsim/python so here goes (btw thanks for this very cool and useful piece of software, I think this is a substantial contribution to the chemical engineering/process discipline):

1) Is it possible to employ COSTALD for liquid density calculation? If so, how can this be specified as an alternative to EOS / EOS volume translated density?
2) How can Kij's/binary interaction parameters for e.g. PR/SRK be user specified/defaults over-wrtitten?
3) What are default Kij's used ? From ChemSep?  
4) Can anyone provide a working example of CCS/sour gas treatment usign e.g. MEA and an electrolyte-activity coeffiecient model (I already asked @EvenSol, so this is for anyone else in the group)

Brds

Anders

**Comment by EvenSol** [comment]
1. COSTALD is not available in NeqSim yet. However this is something that should be added. Do anybody have a reference to a description of Costald? I created a [PR for this task](https://github.com/equinor/neqsim/pull/472). 
2. kij can be displayed with: fluid.getPhase(0).displayInteractionCoefficients("") and you set the kij between compoent 0 and 1 with  fluid.getPhase(0).getMixingRule().setBinaryInteractionParameter(0,1, value) or read it using fluid.getPhase(0).getMixingRule().getBinaryInteractionParameter(0,1)
3. Kij are from NeqSim [own database](https://github.com/equinor/neqsim/wiki/The-NeqSim-parameter-database). The parameters in the default NeqSim database are taken from litterature or fitted to experimental data. Methods for fitting parameters to experimental data is part of NeqSim.

**Comment by andr1976** [comment]
Hi @EvenSol 

Caleb Bell has made an implementation of COSTALD in his python chemicals package. chemicals.volume.COSTALD_mixture(xs, T, Tcs, Vcs, omegas), citing Hankinson, Risdon W., and George H. Thomson. “A New Correlation for Saturated Densities of Liquids and Their Mixtures.” AIChE Journal 25, no. 4 (1979): 653-663. doi:10.1002/aic.690250412.
However, please note that apparently Caleb uses the critical volume instead of the so-called characteristic Volume used by Hankinson and Thomson. The characteristic volume can be calculated from: 

``` python
def V_char_pseudo(TC, Pc, Omega_SRK):
    """
    The function calculates characteristic volume [m3/kmole] based on equation 3.-11.4 in
    "the properties of gases and liquids" 4th ed. Ried, Prausnitz, Poling.
    Aplicable for all hydrocarbons, std. error <1.98 %

    Input: Critical temperature TC [K], critical pressure PC [bar], Omega_SRK [-]
    """
    Gas_C = 8.31451 * 0.00001
    V_Char_PSEUDO = TC * Gas_C / Pc * (0.2851686 - 0.0637911 * Omega_SRK + 0.01379173 * Omega_SRK**2) * 1000
    return V_Char_PSEUDO
```

I think the COSTALD method is also described in chapter  3-11 in "The properties of gases and liquids" 4th ed. Ried, Prausnitz, Poling. I have a copy and can send scans if needed. Another important thing is to do a fall-back volume calculation in case the conditions are above the critical point for which the COSTALD density is not defined.  Also a high-pressure/compressed liquid version is available (also in chemicals package). I can send some naive python code for both the saturated and the compressed version. 

The interesting question is if the COSTALD liquid volume also shall be applied in case of liquid or two-phase volume-flash calculations (e.g. UV-flash) (I think an option to do this could be nice).

**Comment by andr1976** [comment]
For https://github.com/equinor/neqsim/pull/472 I am expecting a liquid density of 643 - 646 kg/m3 (the former being Unisim and the latter being DWSIM and in-house code). Using in-house python code I get 649 kg/m3.

**Reply by EvenSol** [reply]
Thanks Anders. Could you check this @WhisperSv ? I thought the test you made was from an example in the book mentioned above. What was the experimental value?

**Reply by Sviatose** [reply]
Hi, yes, the test was from the book "The properties of gases and liquids, example 5-4, page 5.26". 

With the pressure of 698.47 bar and temperature 344.26 K (ethane 70% decane 30%) **the experimental value was 100.8 cm3/mol**, which corresponds to around 630.95 kg/m3 according to my calculations. The calculation gives the result of  101.38 cm3/mol (or 628.14 kg/m3), so the error between test and experiment is 0.57%. 

**The experimental value for the volume at the bubble point pressure of 53.99 bar is 116.43 cm3/mol** (546.25 kg/m3). The calculation gives 115 cm3/mol (553.036 kg/m3).

**Reply by andr1976** [reply]
@WhisperSv  @EvenSol, the saturated liquid density compares better with my calculations (~550 kg/m3). Maybe the pressure correction is different if you're using the Aalto method? Mine is based on 4th edition of "The properties ..." and I guess yours is 5th edition.

**Reply by Sviatose** [reply]
@andr1976 @EvenSol I have also looked to the example from the 4th edition, example 4-3. With the pressure 345 bar and and temperature 344.26 K experimental value 78.22 cm3/mol (CO2 50.2% and n-butane 49.8%). The calculation gives 74.119 cm3/mol (688.59 kg/m3) while  the calculation result in the example 4-3 is 73.2 cm3/mol. 

Yes, the Aalto method was used that can make a difference and, also,  the critical volume is used instead of the characteristic Volume used by Hankinson and Thomson, though the result is only marginally affected if the characteristic values are replaced with the true values according to the book.

**Reply by andr1976** [reply]
I get 73.9 cm3/mol for the same conditions in my own code and 75.9 cm3/mol in DWSIM. This is a pretty good match. Taking that the saturated liquid volume is very close its proably ok, and the pressures compared are quite extreme.  

I can see that in the java code a slightly different approach is followed for the saturated liquid density not using V_R^0 and V_R^del as the original method it seems it is kind of mixed with the Rackett compressibility also - maybe I am just confused.
